rfd/0116-label-expressions.md

```diff
@@ -0,0 +1,394 @@
+role from ever accessing `production` nodes via access request or even another
+role granting limited access.
+
+Take the following `auditor` role for example:
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: "auditor"
+spec:
+  allow:
+    node_labels:
+      '*': '*'
+    logins: ["auditor"]
+```
+
+If a user `alice` had both roles `[all_except_prod, auditor]` then she would be
+able to access `production` nodes with the `auditor` login and all other nodes
+with either the `root` or `auditor` login.
+But if another user `bob` had both roles `[all_except_prod_legacy, auditor]`
+then he would *not* be able to access any `production` nodes due to the matched
+`deny` rule.
+This is one of the most basic examples, demonstrating the value of a simple
+negative match ([feature request](https://github.com/gravitational/teleport/issues/20145)).
+
+Another [popular ask](https://github.com/gravitational/teleport/issues/10204) is
+for label matchers to support logical "OR" matches, where a node can match any
+one of a set of possibly allowed labels.
+Currently this is not possible without creating multiple roles.
+Label expressions will be able to support this:
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: example
+spec:
+  allow:
+    logins: ["example"]
+
+    # This label expression would grant access to nodes in one of many allowed
+    # environments
+    node_labels_expression: |
+      labels["env"] == "dev" ||
+      labels["env"] == "qa" ||
+      labels["env"] == "staging"
+```
+
+## Details
+
+### Expression Syntax
+
+Every label expression has access to the "evaluation context" as its input and
+must evaluate to a Boolean true/false to indicate a match of the given resource.
+Label expressions can appear in `allow` or `deny` role conditions.
+
+#### Evaluation Context
+
+Label expressions will have access to the following context:
+
+| Syntax | Alias | Type | Description | 
+|--------|-------|------|-------------|
+| `resource.metadata.labels` | `labels` | `map[string]string` | Labels of the resource (node, app, db, etc.) being accessed |
+| `user.spec.traits` | | `map[string][]string` | `external` or `internal` traits of the user accessing the resource |
+
+#### Helper functions
+
+Supported helper functions are a subset of those already supported in role
+templates, `where` expressions, and login rules.
+
+Below, any variable named `list` can contain a list of items (like the list of
+values for a specific user trait) or a single value (like the value of a
+resource label or a string literal).
+
+| Syntax | Return type | Description | Example |
+|--------|-------------|-------------|---------|
+| `contains(list, item)` | Boolean | Returns true if `list` contains an exact match for `item` | `contains(user.spec.traits[teams], labels["team"])` |
+| `regexp.match(list, re)` | Boolean | Returns true if `list` contains a match for `re` | `regexp.match(labels["team"], "dev-team-\d+$")` |
```

Comment:
Regular expressions open us up to potential denial of service risks.  Are we certain regular expressions are a needed feature?  If so, are there any plan to limit the complexity of what patterns are allowed?

Reply:
I did not realize we already accept regular expressions around labels.  It sounds like this is not a new concern, so I will resolve it for this RFD.  I did however file this `low` severity issue with some options for how we can generally improve regex safety: https://github.com/gravitational/security-findings/issues/27

Feel free to respond on that issue with any missing details or other ideas.  Thank you everyone!